File: src/main/java/com/github/zyypj/tadeuBooter/minecraft/tool/text/response/ResponseWaiter.java
package com.github.zyypj.tadeuBooter.minecraft.tool.text.response;

import lombok.Getter;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.player.AsyncPlayerChatEvent;
import org.bukkit.event.player.PlayerQuitEvent;
import org.bukkit.plugin.java.JavaPlugin;

import java.math.BigInteger;
import java.util.HashMap;
import java.util.Map;
import java.util.function.Consumer;

/**
 * Classe responsável por aguardar a resposta do jogador
 * @author syncwrld
 * @website: <a href="https://github.com/syncwrld">...</a>
 */
public class ResponseWaiter implements Listener {

    private final Map<Player, ResponseData> responses = new HashMap<>();

    public void setup(JavaPlugin plugin) {
        plugin.getServer().getPluginManager().registerEvents(this, plugin);
    }

    public void ask(Player player, RequiredType type, Consumer<String> onComplete) {
        responses.put(player, new ResponseData(type, onComplete));
    }

    @EventHandler
    public void onPlayerChat(AsyncPlayerChatEvent event) {
        Player player = event.getPlayer();

        if (!responses.containsKey(player)) return;

        event.setCancelled(true);
        ResponseData responseData = responses.get(player);
        String message = event.getMessage();

        if (message.equalsIgnoreCase("cancelar")) {
            responses.remove(player);
            player.sendMessage("§cAção cancelada!");
            return;
        }

        if (!responseData.getType().isValid(message)) {
            player.sendMessage(responseData.getType().getErrorMessage());
            return;
        }

        responses.remove(player);
        responseData.getOnComplete().accept(message);
    }

    @EventHandler
    public void onPlayerQuit(PlayerQuitEvent event) {
        responses.remove(event.getPlayer());
    }

    @Getter
    private static class ResponseData {
        private final RequiredType type;
        private final Consumer<String> onComplete;

        public ResponseData(RequiredType type, Consumer<String> onComplete) {
            this.type = type;
            this.onComplete = onComplete;
        }
    }

    @Getter
    public enum RequiredType {
        INTEGER("§cVocê deve digitar um número válido."),
        DOUBLE("§cVocê deve digitar um número decimal válido."),
        STRING("§cEntrada inválida."),
        CONFIRMATION("§cVocê deve digitar 'sim' para confirmar ou 'não' para cancelar.");

        private final String errorMessage;

        RequiredType(String errorMessage) {
            this.errorMessage = errorMessage;
        }

        public boolean isValid(String input) {
            try {
                switch (this) {
                    case INTEGER:
                        new BigInteger(input);
                        return true;
                    case DOUBLE:
                        Double.parseDouble(input);
                        return true;
                    case CONFIRMATION:
                        return input.equalsIgnoreCase("sim") || input.equalsIgnoreCase("não");
                    case STRING:
                    default:
                        return !input.trim().isEmpty();
                }
            } catch (NumberFormatException e) {
                return false;
            }
        }
    }
}

File: src/main/java/com/github/zyypj/tadeuBooter/api/minecraft/lang/text/utils/response/ResponseWaiter.kt
package com.github.zyypj.tadeuBooter.api.minecraft.lang.text.utils.response

import com.github.zyypj.tadeuBooter.api.file.YAML
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.player.AsyncPlayerChatEvent
import org.bukkit.event.player.PlayerQuitEvent
import org.bukkit.plugin.java.JavaPlugin
import java.math.BigInteger

class ResponseWaiter(
    plugin: JavaPlugin, yaml: YAML = YAML("config", plugin), basePath: String = "responses"
) : Listener {

    private val messages = ResponseMessages.fromYaml(yaml, basePath)
    private val responses = mutableMapOf<Player, ResponseData>()

    init {
        plugin.server.pluginManager.registerEvents(this, plugin)
    }

    fun ask(player: Player, type: RequiredType, onComplete: (String) -> Unit) {
        responses[player] = ResponseData(type, onComplete)
    }

    @EventHandler
    fun onPlayerChat(event: AsyncPlayerChatEvent) {
        val player = event.player
        val data = responses[player] ?: return

        event.isCancelled = true
        val msg = event.message

        if (msg.equals("cancelar", ignoreCase = true)) {
            responses.remove(player)
            player.sendMessage(messages.cancel)
            return
        }

        if (!data.type.isValid(msg)) {
            player.sendMessage(ResponseMessages.errorFor(data.type, messages))
            return
        }

        responses.remove(player)
        data.onComplete(msg)
    }

    @EventHandler
    fun onPlayerQuit(event: PlayerQuitEvent) {
        responses.remove(event.player)
    }

    private data class ResponseData(
        val type: RequiredType, val onComplete: (String) -> Unit
    )

    enum class RequiredType {
        INTEGER, DOUBLE, STRING, CONFIRMATION;

        fun isValid(input: String): Boolean = try {
            when (this) {
                INTEGER -> {
                    BigInteger(input); true
                }

                DOUBLE -> {
                    input.toDouble(); true
                }

                CONFIRMATION -> input.equals("sim", ignoreCase = true) || input.equals("não", ignoreCase = true)
                STRING -> input.isNotBlank()
            }
        } catch (_: Exception) {
            false
        }
    }
}